server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -123,6 +127,8 @@ protected DocumentParserContext(MappingLookup mappingLookup, MappingParserContex
         this.dynamicObjectMappers = new HashMap<>();
         this.dynamicRuntimeFields = new ArrayList<>();
         this.dimensions = this.indexSettings.getMode().buildDocumentDimensions(this.indexSettings);
+        this.parent = mappingLookup.getMapping().getRoot();
+        this.dynamic = parent.dynamic == null ? ObjectMapper.Dynamic.TRUE : parent.dynamic;
```

Comment:
it is slightly odd that this constructor is clearly tailored to be used only by a specific subclass (RootDocumentParserContext). I wonder if it woudl be worth refactoring these two constructors to add one that takes all the arguments, and have the subclass push all of the arguments while having its logic where it belongs ( I mean the dynamic conditional as well as the getRoot).

Reply:
++, I've reworked the constructors here and pushed the special logic into RootDocumentParserContext